### PR TITLE
Documentation improvments

### PR DESCRIPTION
**Pull request by arm61**
Made a lot of improvements to the documentation. I have tried my best to be concise and documented both public facing API and internals.

### CODE REVIEW

File: orsopy/fileio/base.py [hunk 1]
```diff
@@ -373,28 +409,24 @@ def __post_init__(self):
                 )
 
 
-def _read_header_data(file, validate=False) -> Tuple[dict, list, str]:
+def _read_header_data(file: Union[TextIO, str], validate: bool = False) -> Tuple[dict, list, str]:
```
Comment (andyfaff) [inline orsopy/fileio/base.py:412]: Perhaps the return typing should be: `Tuple[List[Dict], list, str]`

File: orsopy/fileio/data_source.py [hunk 2]
```diff
@@ -4,31 +4,57 @@
...
+    :param facility: Facility where the experiment was performed.
+    :param proposalID: Identifier for experiment at a facility.
+    :param doi: Digital object identifier for the experiment, possibly
+        provided by the facility.
+    """
     title: str
     instrument: str
-    date: datetime.datetime
-    probe: Union[Literal["neutrons", "x-rays"]]
+    start_date: str
```
Comment (andyfaff) [inline orsopy/fileio/data_source.py:31]: Is there a reason why datetime.datetime was removed? Fine to have string, but makes it non-validateable.

File: orsopy/fileio/data_source.py [hunk 3]
```diff
@@ -4,31 +4,57 @@
...
+    probe: str
```
Comment (andyfaff) [inline orsopy/fileio/data_source.py:32]: probe has to either be `neutrons` or `x-rays`, why was the typing changed?

File: orsopy/fileio/data_source.py [hunk 4]
```diff
@@ -61,15 +87,26 @@ class Sample(Header):
...
+    :param polarization: Radiation polarization as one of
+        :code:`'unpolarized'`, :code:`'p'`, :code:`'m'`, :code:`'pp'`,
+        :code:`'pm'`, :code:`'mp'`, :code:`'mm'`, or a
+        :py:class:`orsopy.fileio.base.ValueVector`.
+    :param configuration: Description of the instreument configuration (full
+        polarized/liquid surface/etc).
+    """
     incident_angle: Union[Value, ValueRange]
     wavelength: Union[Value, ValueRange]
-    polarization: Optional[Union[Literal['unpolarized', 'p', 'm', 'mm', 'mp', 'pm', 'pp'],
```
Comment (andyfaff) [inline orsopy/fileio/data_source.py:67]: Why were the Literals removed? It basically means the creator of the file can supply any old string for polarization when we need them to be one of a small set. 
I would've thought we would want the typing for validation, so if we eventually want to generate the schema from this code, then the original typing needs to be kept.

File: orsopy/fileio/data_source.py [hunk 5]
```diff
@@ -82,24 +119,35 @@ class InstrumentSettings(Header):
...
     data_files: List[Union[File, str]]
     references: Optional[List[Union[File, str]]] = None
-    scheme: Optional[Union[
-        Literal[
-            "angle- and energy-dispersive",
-            "angle-dispersive",
-            "energy-dispersive",
-        ]
-    ]] = None
+    scheme: Optional[str] = None
```
Comment (andyfaff) [inline orsopy/fileio/data_source.py:134]: See previous comments on typing (why have the Literals been removed?)

File: orsopy/fileio/reduction.py [hunk 6]
```diff
@@ -12,15 +12,31 @@
...
 
 
 @dataclass(repr=False)
 class Reduction(Header):
-    """A description of the reduction that has been performed."""
+    """
+    A description of the reduction that has been performed.
+
+    :param software: Software used for reduction.
+    :param timestep: Datetime of reduced file creation.
```
Comment (andyfaff) [inline orsopy/fileio/reduction.py:33]: typo, timestamp

File: tests/test_fileio/test_data_source.py [hunk 7]
```diff
@@ -36,26 +35,25 @@ def test_to_yaml(self):
         """
         value = data_source.Experiment('My First Experiment',
                                        'A Lab Instrument',
-                                       datetime(1992, 7, 14, 10, 10,
-                                                10), 'x-rays')
+                                       datetime(1992, 7, 14).strftime("%Y-%m-%d"), 'x-rays')
         assert value.to_yaml() == 'title: My First Experiment\n'\
-            + 'instrument: A Lab Instrument\ndate: 1992-07-14T'\
-            + '10:10:10\nprobe: x-rays\n'
+            + 'instrument: A Lab Instrument\nstart_date: \'1992-07-14\''\
```
Comment (andyfaff) [inline tests/test_fileio/test_data_source.py:40]: IMO the comparison of a string with the output of a `to_yaml` is brittle. If the files validate, and we can check individual entries are correct, then all this test ensures is that the yaml package produces yaml.

File: Makefile [hunk 8]
```diff
@@ -48,7 +48,7 @@ clean-test: ## remove test and coverage artifacts
 	rm -fr .pytest_cache
 
 lint: ## check style with flake8
-	flake8 orsopy tests
+	flake8 --ignore=F401,E501 orsopy tests
```
[the remark this answers is not in the document]
Comment (arm61) [reply Makefile:51]: The `-ignore=E501` means that we don't worry about line length in the `make lint`, which was what @aglavic wanted. Are you suggesting that we change this?

File: orsopy/fileio/data_source.py (hunk 4 above)
Comment (arm61) [reply orsopy/fileio/data_source.py:67]: See above

File: orsopy/fileio/data_source.py (hunk 5 above)
Comment (arm61) [reply orsopy/fileio/data_source.py:134]: See above

File: tests/test_fileio/test_data_source.py (hunk 7 above)
Comment (arm61) [reply tests/test_fileio/test_data_source.py:40]: Suggestions on how we improve?

File: Makefile (hunk 8 above)
[the remark this answers is not in the document]
Comment (andyfaff) [reply Makefile:51]: My bad. I thought it was in the main ci script.

File: tests/test_fileio/test_data_source.py (hunk 7 above)
Comment (andyfaff) [reply tests/test_fileio/test_data_source.py:40]: I'm not sure what the test is examining. It's basically seeing that pyyaml is emitting yaml.

Comment (arm61) [reply tests/test_fileio/test_data_source.py:40]: Is your suggestion then that it is removed?

Comment (bmaranville) [reply tests/test_fileio/test_data_source.py:40]: Can we validate against the schema?

**Comment by arm61** [comment]
@bmaranville @aglavic @jochenstahn @maxskoda any other comments on this PR?

**Pull request merged**